compiler/rustc_target/src/asm/avr.rs

```diff
@@ -0,0 +1,128 @@
+use super::{InlineAsmArch, InlineAsmType};
+use rustc_macros::HashStable_Generic;
+use std::fmt;
+
+def_reg_class! {
+    Avr AvrInlineAsmRegClass {
+        reg,
+        reg_xyz,
+    }
+}
+
+impl AvrInlineAsmRegClass {
+    pub fn valid_modifiers(self, _arch: InlineAsmArch) -> &'static [char] {
+        &[]
+    }
+
+    pub fn suggest_class(self, _arch: InlineAsmArch, _ty: InlineAsmType) -> Option<Self> {
+        None
+    }
+
+    pub fn suggest_modifier(
+        self,
+        _arch: InlineAsmArch,
+        _ty: InlineAsmType,
+    ) -> Option<(char, &'static str)> {
+        None
+    }
+
+    pub fn default_modifier(self, _arch: InlineAsmArch) -> Option<(char, &'static str)> {
+        None
+    }
+
+    pub fn supported_types(
+        self,
+        _arch: InlineAsmArch,
+    ) -> &'static [(InlineAsmType, Option<&'static str>)] {
+        match self {
+            Self::reg => types! { _: I8; },
+            Self::reg_xyz => types! { _: I16; },
+        }
+    }
+}
+
+def_regs! {
+    Avr AvrInlineAsmReg AvrInlineAsmRegClass {
+        r0: reg = ["r0"],
+        r1: reg = ["r1"],
+        r2: reg = ["r2"],
+        r3: reg = ["r3"],
+        r4: reg = ["r4"],
+        r5: reg = ["r5"],
+        r6: reg = ["r6"],
+        r7: reg = ["r7"],
+        r8: reg = ["r8"],
+        r9: reg = ["r9"],
+        r10: reg = ["r10"],
+        r11: reg = ["r11"],
+        r12: reg = ["r12"],
+        r13: reg = ["r13"],
+        r14: reg = ["r14"],
+        r15: reg = ["r15"],
+        r16: reg = ["r16"],
+        r17: reg = ["r17"],
+        r18: reg = ["r18"],
+        r19: reg = ["r19"],
+        r20: reg = ["r20"],
+        r21: reg = ["r21"],
+        r22: reg = ["r22"],
+        r23: reg = ["r23"],
+        r24: reg = ["r24"],
+        r25: reg = ["r25"],
+        r26: reg = ["r26", "XL"],
+        r27: reg = ["r27", "XH"],
+        r28: reg = ["r28", "YL"],
+        r29: reg = ["r29", "YH"],
+        r30: reg = ["r30", "ZL"],
+        r31: reg = ["r31", "ZH"],
```

Comment:
On all other architectures the register names are lowercase. Is this expected to be uppercase here? The compiler is case-sensitive.

Reply:
The "special" registers (`[XYZ][LH]?` and `SP[LH]?` as well as some more exotic "registers") seem to generally be written in uppercase in docs & example code.  But perhaps for consistency it would be preferable to write them lowercase here?  I'm happy to switch that over.

Reply:
I vote to keep `X,Y,Z` registers and other registers mentioned above by Andrew as uppercase for consistency with Atmel docs. The GNU and LLVM assemblers for AVR accepts either case. If it were lowercase-only I think it would surprise users (if it's case sensitive, we should at least put it in the case the're most likely going to use - which my best guess is upper case but who knows with production code)

Reply:
I agree that the copy-paste from the Atmel docs is a key use-case, and so should just work.

We could make both uppercase and lowercase aliases, so that it behaves the same as the assemblers.  But something about that idea seems a bit ugly.

What do you think @Amanieu?

Reply:
I'm not too familiar with AVR so I will leave the decision to you. Just make sure the documentation in the unstable book is consistent with what is actually implemented.